src/kernel/mem/vmm.c

```diff
@@ -3,13 +3,104 @@
 #include <vm.h>
 #include <video.h>
 #include <pmm.h>
-#include <logging.h> 
+#include <logging.h>
+#include <kernel.h>
+#include <bitmap.h>
+#include <util.h>
 
 extern uint64_t p4_table[];
 extern uint64_t p3_table_hh[];
 extern uint64_t p2_table[];
 extern uint64_t pt_tables[];
 
+VmmContainer *vmm_container_root;
+VmmContainer *vmm_cur_container;
+VmmItem *vmm_head;
+
+size_t vmm_items_per_page;
+size_t vmm_cur_index;
+
+//TODO find better naming and probably one of them is not needed
+size_t start_of_vmm_area;
+size_t next_available_address;
+uint64_t end_of_vmm_space;
+/**
+ * When initialized the VM Manager should reserve a portion of the virtual memory space for itself.
+ */
+void vmm_init() {
+
+    vmm_container_root = ((uint64_t) HIGHER_HALF_ADDRESS_OFFSET + VM_KERNEL_MEMORY_PADDING);
+    end_of_vmm_space = (uint64_t) vmm_container_root + VMM_RESERVED_SPACE_SIZE;
+    start_of_vmm_area = (size_t) vmm_container_root + VMM_RESERVED_SPACE_SIZE + VM_KERNEL_MEMORY_PADDING;
+    next_available_address = start_of_vmm_area;
+    vmm_items_per_page = (PAGE_SIZE_IN_BYTES / sizeof(VmmItem)) - 1;
+    vmm_cur_index = 0;
+
+    loglinef(Verbose, "(vmm_init) Vmm root comparison: (vmm_root) %x - %x (end_of_vmm_space)", vmm_container_root, end_of_vmm_space);
+    //I need to compute the size of the VMM address space
+    uint64_t vmm_root_phys = pmm_alloc_frame();
+    if (vmm_root_phys == NULL) {
+        loglinef(Verbose, "(vmm_init)  vmm_root_phys should not be null");
+        return;
+    }
+    
+    loglinef(Verbose, "(vmm_init) Got vmm_root_phys address: %x", vmm_root_phys); 
+    // Mapping the phyiscal address for the vmm structures
+    map_phys_to_virt_addr(vmm_root_phys, vmm_container_root, 0);    
+    loglinef(Verbose, "(vmm_init) Testing the just mapped address");
+    vmm_container_root->vmm_root[0].base = 0;
+    vmm_container_root->vmm_root[0].size = 5;
+    vmm_container_root->vmm_root[0].flags = 0;
+    loglinef(Verbose, "(vmm_init) flags should be 0: %d size should be 5: %d", vmm_container_root->vmm_root[0].flags, vmm_container_root->vmm_root[0].size);
+    loglinef(Verbose, "(vmm_init) where does the container  start? %x", &vmm_container_root);
+    loglinef(Verbose, "(vmm_init) where does the next end? %x", &(vmm_container_root->next));
+    loglinef(Verbose, "(vmm_init) start of vmm_area %x", start_of_vmm_area);
+    vmm_container_root->next = NULL;
+    vmm_cur_container = vmm_container_root;
+    vmm_head = NULL;
+}
+
+void *vmm_alloc(size_t length, size_t flags) {
+
+    //TODO When the space inside this page is finished we need to allocate a new page
+    //     at vmm_cur_container + sizeof(VmmItem)
+    if (length < 0) {
+        return NULL;
+    }
+
+    if ( vmm_head + sizeof(VmmItem) + length > end_of_vmm_space ) {
```

Comment:
pointer maths haha! vmm_head is `VmmItem*`, not `uint8_t*`.

Reply:
that if should not be there anymore